lib/awful/widget/inputbox.lua

```diff
@@ -0,0 +1,530 @@
+                    self._private.text = text_start .. text_end
+                    self._private.highlight = {}
+                    cursor_pos = #text_start + 1
+                else
+                    if cursor_pos <= #self._private.text then
+                        self._private.text = self._private.text:sub(1, cursor_pos - 1) ..
+                            self._private.text:sub(cursor_pos + 1)
+                    end
+                end
+            elseif key == "Left" then
+                -- Move cursor ro the left
+                if cursor_pos > 1 then
+                    cursor_pos = cursor_pos - 1
+                end
+                self._private.highlight = {}
+            elseif key == "Right" then
+                -- Move cursor to the right
+                if cursor_pos <= #self._private.text then
+                    cursor_pos = cursor_pos + 1
+                end
+                self._private.highlight = {}
+            else
+                -- Print every alphanumeric key
+                -- It seems like gears.xmlescape doesn't support non alphanumeric characters
+                if key:wlen() == 1 then
+                    self._private.text = self._private.text:sub(1, cursor_pos - 1) ..
+                        key .. self._private.text:sub(cursor_pos)
+                    cursor_pos = cursor_pos + 1
+                end
+            end
+
+            -- Make sure the cursor cannot go out of bounds
+            if cursor_pos < 1 then
+                cursor_pos = 1
+            elseif cursor_pos > #self._private.text + 1 then
+                cursor_pos = #self._private.text + 1
+            end
+        end
+
+        -- Update cycle
+        self:update(self._private.text, cursor_pos)
+
+        if callbacks.pressed then
+            callbacks.pressed(key, mod_keys)
+        end
+    end)
+end
+
+--- Creates a new inpubox widget
+-- @tparam table args Arguments for the inpubox widget
+-- @tparam string args.text The text to display in the inpubox
+-- @tparam[opt=beautiful.fg_normal] string args.fg Text foreground color
+-- @tparam[opt=beautiful.border_focus] string args.border_focus_color Border color when focused
+-- @tparam[opt=""] string args.hint_text Hint text to be shown when not focused and
+-- @tparam[opt=beautiful.inputbox_hint_fg] string args.hint_fg Hint text foreground color
+-- @tparam[opt=beautiful.inputbox_cursor_bg] string args.cursor_bg Cursor background color
+-- @tparam[opt=beautiful.inputbox_cursor_fg] string args.cursor_fg Cursor foreground color
+-- @tparam[opt=beautiful.inputbox_highlight_bg] string args.highlight_bg Highlight background color
+-- @tparam[opt=beautiful.inputbox_highlight_fg] string args.highlight_fg Highlight foreground color
+-- @treturn awful.widget.inputbox The inputbox widget.
+-- @constructorfct awful.widget.inputbox
+function inputbox.new(args)
+    args = args or {}
+
+    local ret = gobject { enable_properties = true }
+    gtable.crush(ret, inputbox, true)
```

Comment:
This object is not returned from the constructor, so the methods assigned to `ret` and the signals emitted from it won't be accessible to the user.

You have two options:
- Assign your methods to the actual widget. Downside is you need to do extra stuff to hide the widget's own API, and it might get complicated with a template
- Return `ret` from the constructor and implement `.fit` and `:layout` methods to make `inputbox` itself a container that renders the inner textbox

Reply:
I tired to implement that with #3421 but I couldn't figure out how exactly since there is now wibox.widget.template in place of a base widget.